### PR TITLE
Tbox-6-Refactor-Notifications

### PR DESCRIPTION
**Pull request by OGordon100**
# ↪️ Pull Request

Right now in TBOX 61 and 62, everything is coded as semi-duplicate functions (e.g. notifications.common._monitor_job being called by notifications.slack.monitor_job and notifications.emails.monitor_job, with each monitor_job() implementing everything related to sending the notification, and not just monitoring the job.)

This not only means have large amounts of docs dedicated to utility functions that immediately return other functions, but we end up making increasingly duplicated code for:
a) More notification types (e.g. Teams)
b) Notifying for new tasks not related to monitoring jobs (e.g. reminders about new assignments). Already there is some functionality currently coded for emails but not for slack)

(NB here - in presales we very often end up discussing means of notifying pair assignments, with emails, slack, teams, etc, a frequent request that is so frequent it should probably be in the tbox)

If we are willing to deprecate, could I propose a class based solution to this problem? 

We would have a _BaseNotifier(object) class that implements the current monitor_job (without email/slack specific arguments attached to it, This itself would call a send_message(text, *args, **kwargs) func (that in the BaseNotifier returns NotImplementedError).

A slack/email/etc specific EmailNotifier(sender_address, sender_password, ...), etc, class, would then implement its own email, etc, specific send_message(text, subject_line, recipient_addresses, ...) function for sending the specific message, job monitoring or otherwise. This is also more easily filled from a yaml file.

We could even have a bit of extra code for email-specific parsing of the response for formatting, by having an EmailNotifiers own monitor_job(recipient_addresses) , which calls super().monitor_job(), then does email specific stuff to the output.

Logistically it is a little more complicated than the current separate definitions implementation (core logic in the BaseNotifier, and formatting/calling specific in the type specific class), but once you have one, you have an obvious “template” to go off. I’ve done this type of thing several times in the past to good effect, and I think notifications is a perfect candidate.

Not only would Slack, Teams, etc, build off this core logic for triggering notifications making development of future notification means easier, but we could also add other functions to the BaseNotifier (such as assignment notifications), and all means of notification would automatically be able to deal with them, as they are then just sending a different message.  

**This ticket deprecates the old notification functionality and refactors it into a class based structure. It also implements `TeamsNotifier`. It is also a requirement for [TBOX-13
](https://datatamr.atlassian.net/jira/software/c/projects/TBOX/boards/107?modal=detail&selectedIssue=TBOX-13), as the TBEX implementation will extend `_BaseNotifier` to include `monitor_assignments()`**

## ✔️ PR Todo

- [x] Add documentation
- [x] Add examples
- [x] Ensure you are following the practices in the [contributor guide](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit) and [coding_standards](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards)
  * [Use consistent and descriptive name conventions](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit#heading=h.477s36w4rds9) 
  * [Uses google-style docstrings](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Google-style-docstrings)
  * [Uses versioned API/client methods only](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit#heading=h.yv6df99fyrq2)
  * [Avoid global variables](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Avoid-global-variables)
  * [Avoid giant try/catch](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Avoid-giant-try-/-except)
  * [Don't use mutable default variables](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Don%E2%80%99t-use-mutable-default-arguments)
  * [Use a `main()` function in your scripts](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Use-a-%E2%80%9Cmain()%E2%80%9D--function-in-your-scripts)
- [x] Added/updated testing for this change (ensure you are testing any changes/new code!)
- [x] Update title to link to TBOX issue (i.e. start pull request title with TBOX-xx) 
- [x] Approval from first code-reviewer
- [ ] Approval from second code-reviewer (only required if a second reviewer has left comments)
- [x] Pass all checks

### CODE REVIEW

**Comment by OGordon100** [comment]
Commit #[27bdc38](https://github.com/Datatamer/tamr-toolbox/pull/109/commits/27bdc3834b87b60a1b8b083c0fceb1ff90d7ac31) is suitable for an initial release of TBOX. 

Per-user notifications are not possible, however (planned for by extending the recipients input to be a dict linking Tamr usernames to destinations), and it would be beneficial to hold off a merge for a little bit longer.

File: examples/snippets/notifications/enable_teams_notifications.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+"""Example script for generating Slack notifications based on Tamr jobs"""
```
Comment (Kovacsova) [inline examples/snippets/notifications/enable_teams_notifications.py]: Typo; "Teams" notification

File: tamr_toolbox/notifications/common.py [hunk 2]
```diff
@@ -1,12 +1,123 @@
 """Tasks related to creation of Email notifications"""
```
Comment (Kovacsova) [inline tamr_toolbox/notifications/common.py]: All notifications?

File: tamr_toolbox/notifications/slack.py [hunk 3]
```diff
@@ -1,24 +1,71 @@
-"""Tasks related to creation of Slack notifications"""
 import logging
-import os
+import warnings
+from abc import ABC
+from dataclasses import dataclass
+from typing import Dict
```
Comment (Kovacsova) [inline tamr_toolbox/notifications/slack.py]: can be on one line

File: tamr_toolbox/notifications/slack.py [hunk 4]
```diff
@@ -1,24 +1,71 @@
...
+        recipients = self._parse_recipients(tamr_user)
+
+        for recipient in recipients:
+            LOGGER.info(f"Sending a Slack message to {recipient}")
+            try:
+                self.slack.chat_postMessage(
+                    channel=recipient,
+                    text=message,
+                    username="Tamr Notifications Bot",
+                    icon_url="https://jdp491bprdv1ar3uk2puw37i-wpengine.netdna-ssl.com/wp-content/uploads/2020/08/Tamr-Square-Dark.png",  # noqa
```
Comment (Kovacsova) [inline tamr_toolbox/notifications/slack.py]: add comment

File: tamr_toolbox/notifications/emails.py [hunk 5]
```diff
@@ -1,22 +1,100 @@
...
+    smtp_server: str
+    smtp_port: int
+    use_tls: Optional[bool] = False
+    keyfile: Optional[str] = None
+    certfile: Optional[str] = None
+    """
+    Send emails based on Tamr eventing.
+
+    Args:
+        recipients:
```
Comment (Kovacsova) [inline tamr_toolbox/notifications/emails.py]: rename to email_recipients

File: tamr_toolbox/notifications/common.py [hunk 6]
```diff
@@ -1,12 +1,123 @@
...
+                    f"than {timeout} seconds to resolve."
+                )
+                self.send_message(message=timeout_message, title=f"Job {op.resource_id}: Timeout")
+
+    def _parse_recipients(self, tamr_user: Optional[str] = None) -> List[str]:
+        """Look up Tamr users & parse recipients into an iterable list"""
+        if type(self.recipients) in [str, list]:
+            if tamr_user:
+                raise ValueError(
+                    "If specifying a Tamr user, the recipients class input"
```
Comment (Kovacsova) [inline tamr_toolbox/notifications/common.py:106]: could you use this in one example please?

File: tamr_toolbox/utils/operation.py [hunk 7]
```diff
@@ -49,7 +49,7 @@ def get_latest(tamr: Client) -> Operation:
     return op
 
 
-def get_details(*, operation: Operation) -> str:
+def get_details(operation: Operation) -> str:
```
Comment (Kovacsova) [inline tamr_toolbox/utils/operation.py]: please revert

**Review by Kovacsova — COMMENTED** [review]
This is amazing, see small comments

File: tamr_toolbox/notifications/common.py (hunk 6 above)
Comment (OGordon100) [reply tamr_toolbox/notifications/common.py:106]: Done.

Slack notification example is setup as a string recipient.
Teams notification example is setup as a list of strings recipients
Email notification example is setup as a dict of tamr users:emails recipients

**Review by skalish — CHANGES_REQUESTED** [review]
It looks like the code in this PR currently goes against our functional programming paradigm.  We should not be adding classes with methods to the codebase.  Dataclasses are okay, but all operations need to be written as functions and not methods.

**Comment by OGordon100** [comment]
Moved to TBEX 

Shall be an interesting comparison of functional vs object usage!

https://github.com/Datatamer/tamr-toolbox-extensions/pull/11

**Pull request closed without merging**